pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/ConcurrentMapPartitionUpsertMetadataManager.java

```diff
@@ -74,6 +75,16 @@ protected void addOrReplaceSegment(ImmutableSegmentImpl segment, ThreadSafeMutab
       @Nullable ThreadSafeMutableRoaringBitmap queryableDocIds, Iterator<RecordInfo> recordInfoIterator,
       @Nullable IndexSegment oldSegment, @Nullable MutableRoaringBitmap validDocIdsForOldSegment) {
     String segmentName = segment.getSegmentName();
+    // Skip adding segments that has segment EndTime in the comparison cols earlier than (largestSeenTimestamp - TTL).
+    // Note: We only update largestSeenComparisonValueMs when addRecord, and access the value when addSegments.
+    // We only support single comparison column for TTL-enabled upsert tables.
+    if (_largestSeenComparisonValue > 0) {
```

Comment:
Can we move this logic into the base manager?

Reply:
sure. i added this to both addSegment and replaceSegment.